pkg/shardddl/optimism/operation.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optimism
+
+import (
+	"context"
+	"encoding/json"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/clientv3util"
+	"go.etcd.io/etcd/mvcc/mvccpb"
+
+	"github.com/pingcap/dm/dm/common"
+	"github.com/pingcap/dm/pkg/etcdutil"
+)
+
+// ConflictStage represents the current shard DDL conflict stage in the optimistic mode.
+type ConflictStage string
+
+const (
+	// ConflictNone indicates no conflict exists,
+	// DM-worker can execute DDL/DML to the downstream normally.
+	ConflictNone ConflictStage = "none"
+	// ConflictDetected indicates a conflict will exist after applied the shard DDL.
+	// in this stage, DM-worker should not execute/skip DDL/DML,
+	// but it should still try to find the DDL which can resolve the conflict in the binlog stream.
+	ConflictDetected ConflictStage = "detected"
+	// ConflictResolved indicates a conflict will be resolved after applied the shard DDL.
+	// in this stage, DM-worker should replay DML skipped in ConflictDetected to downstream.
+	ConflictResolved ConflictStage = "resolved"
+)
+
+// Operation represents a shard DDL coordinate operation.
+// This information should be persistent in etcd so can be retrieved after the DM-master leader restarted or changed.
+// NOTE: `Task`, `Source`, `UpSchema` and `UpTable` are redundant in the etcd key path for convenient.
+// Operation is putted when coordinating a shard DDL operation for DM-worker by DM-master,
+// and is updated (with `done`) after DM-worker has done the operation by DM-worker,
+// and is deleted when removing the lock by DM-master.
+// because we need the newest stage in Operation to recover the lock when restarting DM-master.
+type Operation struct {
+	ID            string        `json:"id"`             // the corresponding DDL lock ID
+	Task          string        `json:"task"`           // data migration task name
+	Source        string        `json:"source"`         // upstream source ID
+	UpSchema      string        `json:"up-schema"`      // upstream/source schema name, different sources can have the same schema name
+	UpTable       string        `json:"up-table"`       // upstream/source table name, different sources can have the same table name
+	DDLs          []string      `json:"ddls"`           // DDL statements need to apply to the downstream.
```

Comment:
It sames that the only usage of `DDLs` in this PR is to check its length? What will we do with this field in the future?

Reply:
It's my mistake. _syncer_ should replicate these `DDLs` to downstream, I updated in https://github.com/pingcap/dm/pull/568/commits/6c8c1177b5bdb81164b21e050e690596be3306d4.